internal/build/icann.go

```diff
@@ -0,0 +1,176 @@
+		return err
+	}
+	defer res.Body.Close()
+
+	var x icannGTLDResponse
+	d := json.NewDecoder(res.Body)
+	err = d.Decode(&x)
+	if err != nil {
+		return err
+	}
+
+	if x.Version != 2 {
+		return fmt.Errorf("unknown ICANN gTLDs list version: %d", x.Version)
+	}
+
+	var tagsAdded, tagsRemoved, zonesWithdrawn, zonesRetired, zonesModified int
+	for _, g := range x.GTLDs {
+		domain := Normalize(g.GTLD)
+		z, ok := zones[domain]
+		if !ok {
+			if len(zones) > 100 {
+				Trace("@{y}Unknown domain in ICANN gTLD feed: %s\n", domain)
+			}
+			continue
+		}
+
+		var modified bool
+
+		switch {
+		// gTLD was withdrawn
+		case g.DelegationDate.IsZero() && g.ContractTerminated:
+			z.AddTags(TagWithdrawn)
+			z.RemoveTags(TagRetired)
+			zonesWithdrawn++
+			modified = true
+
+		// gTLD was retired
+		case !g.RemovalDate.IsZero():
+			z.RemoveTags(TagWithdrawn)
+			z.AddTags(TagRetired)
+			zonesRetired++
+			modified = true
+
+		// Pending delegation
+		case g.DelegationDate.IsZero() && !g.ContractTerminated:
+			n := len(z.NameServers)
+			if n > 0 {
+				Trace("@{y}Warning: undelegated gTLD with %d name servers: @{y!}%s\n", n, z.Domain)
+			}
+		}
+
+		// Brand TLD?
+		if g.Specification13 {
+			d := z.AddTags(TagBrand)
+			if d != 0 {
+				tagsAdded += d
+				modified = true
+			}
+		} else if !z.IsRetiredOrWithdrawn() && z.IsBrand() {
+			// https://newgtlds.icann.org/en/applicants/agb/base-agreement-contracting/specification-13-applications
+			Trace("@{y}Brand gTLD without ICANN Specification 13: @{y!}%s\n", z.Domain)
+		}
+
+		// Extract registry operator
+		if g.RegistryOperator != "" {
+			z.RegistryOperator = g.RegistryOperator
+			modified = true
+		}
+
+		if modified {
+			zonesModified++
+		}
+	}
+
+	Trace("@{.}Added %d tag(s), removed %d tag(s) from %d zone(s)\n", tagsAdded, tagsRemoved, zonesModified)
+	Trace("@{.}Widthdrew %d and retired %d zone(s)\n", zonesWithdrawn, zonesRetired)
+
+	return nil
+}
+
+type icannGTLDResponse struct {
+	GTLDs []struct {
+		ApplicationID                 string      `json:"applicationID"`
+		ContractTerminated            bool        `json:"contractTerminated"`
+		DateOfContractSignature       ISODate     `json:"dateOfContractSignature"`
+		DelegationDate                ISODate     `json:"delegationDate"`
+		GTLD                          string      `json:"gTLD"`
+		RegistryClassDomainNameList   interface{} `json:"registryClassDomainNameList"` // (always null)
+		RegistryOperator              string      `json:"registryOperator"`
+		RegistryOperatorCountryCode   *string     `json:"registryOperatorCountryCode"` // (always null)
+		RemovalDate                   ISODate     `json:"removalDate"`
+		Specification13               bool        `json:"specification13"`               // Brand TLD
+		ThirdOrLowerLevelRegistration bool        `json:"thirdOrLowerLevelRegistration"` // (always false or null)
+		ULabel                        string      `json:"uLabel"`                        // Unicode IDN label
```

Comment:
Do you see these in the real feed? Where might we use them, if anywhere?

Reply:
Good question. Maybe to validate our Unicode normalization?

Reply:
Should we... write a test about it? I see the warning message that was added. Anyone reviewing those terminal outputs?